File: 2023/Leetcode-/problem200.py
class Solution:
    def __init__(self):
        self.WATER = "0"
        self.ISLAND = "1"

    def findIsland(self, grid: List[List[str]], i: int, j: int) -> None:
        if 0 <= i < len(grid) and 0 <= j < len(grid[i]) and grid[i][j] == self.ISLAND:
            grid[i][j] = self.WATER
            self.findIsland(grid, i + 1, j)
            self.findIsland(grid, i - 1, j)
            self.findIsland(grid, i, j + 1)
            self.findIsland(grid, i, j - 1)


    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == self.ISLAND:
                    count += 1
                    self.findIsland(grid, i, j)

        return count

File: 2023/Leetcode/problem200.java
class Solution {
    public final char WATER = '0';
    public final char LAND = '1';

    public void dfsTraversal(char[][] grid, int area, int i, int j) {
        grid[i][j] = WATER;

        if (j + 1 < grid[i].length && grid[i][j + 1] == LAND) {
            dfsTraversal(grid, area + 1, i, j + 1);
        }
        if (j - 1 >= 0 && grid[i][j - 1] == LAND) {
            dfsTraversal(grid, area + 1, i, j - 1);
        }
        if (i + 1 < grid.length && grid[i + 1][j] == LAND) {
            dfsTraversal(grid, area + 1, i + 1, j);
        }
        if (i - 1 >= 0 && grid[i - 1][j] == LAND) {
            dfsTraversal(grid, area + 1, i - 1, j);
        }
    }

    public int numIslands(char[][] grid) {
        int islandCount = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] == LAND) {
                    dfsTraversal(grid, 1, i, j);
                    islandCount++;
                }
            }
        }

        return islandCount;
    }
}